internal/langserver/handlers/did_change_watched_files.go

```diff
@@ -0,0 +1,79 @@
+package handlers
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+
+	"github.com/creachadair/jrpc2"
+	"github.com/hashicorp/terraform-ls/internal/job"
+	ilsp "github.com/hashicorp/terraform-ls/internal/lsp"
+	"github.com/hashicorp/terraform-ls/internal/protocol"
+	lsp "github.com/hashicorp/terraform-ls/internal/protocol"
+	"github.com/hashicorp/terraform-ls/internal/terraform/ast"
+)
+
+func (svc *service) DidChangeWatchedFiles(ctx context.Context, params lsp.DidChangeWatchedFilesParams) error {
+	var ids job.IDs
+
+	for _, change := range params.Changes {
+		uri := string(change.URI)
+		fi, err := os.Stat(uri)
+		if err != nil {
+			jrpc2.ServerFromContext(ctx).Notify(ctx, "window/showMessage", &lsp.ShowMessageParams{
+				Type:    lsp.Warning,
+				Message: fmt.Sprintf("Unable to update: %s, Failed to open directory", uri),
+			})
+			continue
+		}
+
+		if !fi.IsDir() {
+			jrpc2.ServerFromContext(ctx).Notify(ctx, "window/showMessage", &lsp.ShowMessageParams{
+				Type:    lsp.Warning,
+				Message: fmt.Sprintf("No such directory for URI: %s", uri),
+			})
+			continue
+		}
+
+		filename := path.Base(uri)
+		if !ast.IsModuleFilename(filename) && !ast.IsVarsFilename(filename) {
+			jrpc2.ServerFromContext(ctx).Notify(ctx, "window/showMessage", &lsp.ShowMessageParams{
+				Type:    lsp.Warning,
+				Message: fmt.Sprintf("Unable to update file: %s, filetype not supported", uri),
+			})
+			continue
+		}
+
+		// only handle `Changed` event type
+		if change.Type == protocol.Changed {
+			dh := ilsp.HandleFromDocumentURI(change.URI)
+
+			// check existence
+			_, err = svc.modStore.ModuleByPath(dh.Dir.Path())
+			if err != nil {
+				return err
+			}
+
+			jobIds, err := svc.parseAndDecodeModule(dh.Dir)
+			if err != nil {
+				return err
+			}
+
+			ids = append(ids, jobIds...)
+
+		}
+
+	}
+
+	// range over all ids corresponding
+	// to all changes in current request
+	for _, id := range ids {
+		err := svc.stateStore.JobStore.WaitForJobs(ctx, id)
```

Comment:
Just curious, since we're "waiting" for jobs here, and given `go-memdb` is concurrent-safe for read ops, is it possible for us to do wait for jobs concurrently? Again, just a thought, I'm not sure how the internal semantics are just yet and I don't think there'll be any considerable difference in perceived performance/experience given we don't usually expect a lot of changes as part of a single request (or do we?).

Reply:
You can replace the for loop entirely with just the call, as the 2nd argument is variadic, i.e.

```go
err := svc.stateStore.JobStore.WaitForJobs(ctx, ids...)
if err != nil {
	return err
}
```

[It will loop over all IDs and wait for each](https://github.com/hashicorp/terraform-ls/blob/16c915b9fb13fb5061fc1a6c71959e3d9421af18/internal/state/jobs.go#L249-L256). There is no benefit in waiting concurrently since we're waiting for the slowest one anyway and the ones which are already done by the time we get to them will just return straight away.

Reply:
Thanks for the clarification, waiting for the slowest one sounds right.